### PR TITLE
Fix Embedded Kafka not delaying topic creation

### PR DESCRIPTION
**Pull request by bcarter97**

### CODE REVIEW

File: src/test/scala/utils/EmbeddedKafka.scala [hunk 1]
```diff
@@ -11,15 +11,22 @@ import uk.sky.fs2.kafka.topicloader.TopicLoader.topicPartitionOrder
...
 
   def createCustomTopic(topic: String, partitions: Int, topicConfig: Map[String, String])(implicit
       kafkaConfig: EmbeddedKafkaConfig,
       F: Async[F]
   ): F[NonEmptyList[TopicPartition]] =
     for {
-      _ <- F.fromTry(Underlying.createCustomTopic(topic = topic, topicConfig = topicConfig, partitions = partitions))
+      maybeCreate <-
+        F.delay(Underlying.createCustomTopic(topic = topic, topicConfig = topicConfig, partitions = partitions))
+      _           <- F.fromTry(maybeCreate)
```
Comment (bcarter97) [inline src/test/scala/utils/EmbeddedKafka.scala:29]: This is the main change. Previously we were wrapping `createCustomTopic`, a Try in `F.fromTry` which wasn't deferring the execution.

File: it/src/test/scala/integration/LoadExampleIntSpec.scala [hunk 2]
```diff
@@ -80,9 +79,15 @@ class LoadExampleIntSpec extends KafkaSpecBase[IO] {
...
-    object testContext extends TestContext[IO]
-    import testContext.*
-    embeddedKafka.surround(test(testContext))
-  }
+  private def withKafkaContext(test: TestContext[IO] => IO[Assertion]): IO[Assertion] =
+    for {
+      config      <- embeddedKafkaConfigF
+      testContext <- new TestContext[IO] {
+                       override implicit val kafkaConfig: EmbeddedKafkaConfig = config
+                     }.pure[IO]
```
Comment (peter-hazell) [inline it/src/test/scala/integration/LoadExampleIntSpec.scala:87]:
```suggestion
      testContext = new TestContext[IO] {
                       override implicit val kafkaConfig: EmbeddedKafkaConfig = config
                     }
```

File: src/test/scala/integration/TopicLoaderIntSpec.scala [hunk 3]
```diff
@@ -214,13 +214,18 @@ class TopicLoaderIntSpec extends KafkaSpecBase[IO] {
...
-    object testContext extends TestContext[IO]
-    import testContext.*
-    embeddedKafka.surround(test(testContext))
-  }
+  private def withKafkaContext(test: TestContext[IO] => IO[Assertion]): IO[Assertion] =
+    for {
+      config      <- embeddedKafkaConfigF
+      testContext <- new TestContext[IO] {
+                       override implicit val kafkaConfig: EmbeddedKafkaConfig = config
+                     }.pure[IO]
```
Comment (peter-hazell) [inline src/test/scala/integration/TopicLoaderIntSpec.scala:225]:
```suggestion
      testContext = new TestContext[IO] {
                       override implicit val kafkaConfig: EmbeddedKafkaConfig = config
                     }
```

File: src/test/scala/integration/TopicLoaderIntSpec.scala [hunk 4]
```diff
@@ -214,13 +214,18 @@ class TopicLoaderIntSpec extends KafkaSpecBase[IO] {
   }
 
   private trait TestContext[F[_]] {
-    implicit val kafkaConfig: EmbeddedKafkaConfig =
-      EmbeddedKafkaConfig(kafkaPort = RandomPort(), zooKeeperPort = RandomPort(), Map("log.roll.ms" -> "10"))
+    implicit val kafkaConfig: EmbeddedKafkaConfig
```
Comment (peter-hazell) [inline src/test/scala/integration/TopicLoaderIntSpec.scala:217]:
```suggestion
    given kafkaConfig: EmbeddedKafkaConfig
```

File: it/src/test/scala/integration/LoadExampleIntSpec.scala [hunk 5]
```diff
@@ -46,20 +45,20 @@ class LoadExampleIntSpec extends KafkaSpecBase[IO] {
   }
 
   private abstract class TestContext[F[_] : Async] {
+
+    implicit val kafkaConfig: EmbeddedKafkaConfig
```
Comment (peter-hazell) [inline it/src/test/scala/integration/LoadExampleIntSpec.scala:49]:
```suggestion
    given kafkaConfig: EmbeddedKafkaConfig
```

File: src/test/scala/integration/TopicLoaderIntSpec.scala (hunk 4 above)
Comment (bcarter97) [reply src/test/scala/integration/TopicLoaderIntSpec.scala:217]: Atm we're cross compiling Scala 2.13 and 3, so we need to use implicits etc. Raised https://github.com/sky-uk/fs2-kafka-topic-loader/issues/9 to discuss dropping Scala 2.13 support as I'm not sure it's necessary.

File: it/src/test/scala/integration/LoadExampleIntSpec.scala [hunk 6]
```diff
@@ -46,20 +45,20 @@ class LoadExampleIntSpec extends KafkaSpecBase[IO] {
...
+
     private val store: F[Ref[F, List[String]]] = Ref[F].of(List.empty)
 
     private implicit val loggerFactory: LoggerFactory[F] = Slf4jFactory.create[F]
```
Comment (peter-hazell) [inline it/src/test/scala/integration/LoadExampleIntSpec.scala:53]: not this PR but anywhere we are using `implicit` shall we replace with `given` ?

Comment (bcarter97) [reply it/src/test/scala/integration/LoadExampleIntSpec.scala:53]: https://github.com/sky-uk/fs2-kafka-topic-loader/pull/13#discussion_r1321340785

File: src/test/scala/integration/TopicLoaderIntSpec.scala (hunk 4 above)
Comment (bcarter97) [reply src/test/scala/integration/TopicLoaderIntSpec.scala:217]: https://github.com/sky-uk/fs2-kafka-topic-loader/blob/9e60e7e924735ebe4c4532f2422b7d556e18eecf/build.sbt#L5

**Review by peter-hazell — APPROVED** [review]

**Review by peter-hazell — APPROVED** [review]

**Review by lacarvalho91 — APPROVED** [review]

**Pull request merged**